register-machines/src/main.rs

```diff
@@ -13,66 +15,165 @@ enum Instruction {
+                let v = rs.entry(i).or_insert(Godel::new()); 
+                if *v != 0 {
+                    *v -= 1;
+                    l = j;
+                } else {
+                    l = k; 
+                }
+            },
+            Halt => { break; }
+        }
+    }
+    (l, rs)
+}
+fn encode_instruction(ins: &Instruction) -> Godel {
+    match ins {
+        Add(i, j) => 
+            encode_pair1(&Godel::from(2 * i), &Godel::from(*j)),
+        Sub(i, j, k) => 
+            encode_pair1(&Godel::from(2 * i + 1),
+            &encode_pair2(&Godel::from(*j), &Godel::from(*k))
+        ),
+        Halt => Godel::ZERO
+    }
 }
 // <<x,y>> = (2^x)*(2y+1)
-fn encode_pair1(x: u128, y: u128) -> u128 {
-    unimplemented!();
+fn encode_pair1(x: &Godel, y: &Godel) -> Godel {
+    (Godel::from(2) * y + 1) << x.to_u32_wrapping()
 }
 // <x,y> = (2^x)*(2y+1)-1
-fn encode_pair2(x: u128, y: u128) -> u128 {
-    unimplemented!();
+fn encode_pair2(x: &Godel, y: &Godel) -> Godel {
+    encode_pair1(x, y) - 1
 }
-fn encode_list_to_godel(l: &[u128]) -> u128 {
-    unimplemented!();
+fn encode_list_to_godel(l: &[Godel]) -> Godel {
+    if l.is_empty() { return Godel::ZERO; } 
+    encode_pair1(&l[0], &encode_list_to_godel(&l[1..])) 
 }
-fn encode_program_to_list(program: &[Instruction]) -> Vec<u128> {
-    unimplemented!();
+fn encode_program_to_list(program: &[Instruction]) -> Vec<Godel> {
+    program.iter().map(|x| encode_instruction(x)).collect()
+}
+// Returns 0 if there are no trailing zeros, else return trailing zero count (in binary)
+fn trailing_zeros_in_binary(x: &Godel) -> Godel {
+    let mut b = x.clone(); 
+    let mut c = Godel::new();
+    if b != 0 {
+        b = (b.clone() ^ (b - 1)) >> 1; 
+        while b != 0 {
+            c += 1;
+            b >>= 1;
+        }
+    }
+    c
+}
+fn decode_instruction(ins: &Godel) -> Instruction {
+    if *ins == 0 { return Halt; }
+    let (x, y) = decode_pair1(ins); 
+    let i: Godel = x.clone() / 2;
+    if x % 2 != 0 {
+        let (j, k) = decode_pair2(&y);
+        Sub(i.try_into().unwrap(), j.try_into().unwrap(), k.try_into().unwrap())
+    } else {
+        Add(i.try_into().unwrap(), y.try_into().unwrap())
+    }
 }
 // a = (2^x)*(2y+1)
-fn decode_pair1(a: u128) -> (u128, u128) {
-    unimplemented!();
+fn decode_pair1(a: &Godel) -> (Godel, Godel) {
+    let x: Godel = trailing_zeros_in_binary(a); 
+    let z = Godel::from(a >> x.to_u32_wrapping());
+    let y: Godel = (z - 1) / 2;
+    (x, y)
 }
 // a = (2^x)*(2y+1)-1
-fn decode_pair2(a: u128) -> (u128, u128) {
-    unimplemented!();
+fn decode_pair2(a: &Godel) -> (Godel, Godel) {
+    decode_pair1(&Godel::from(a + 1))
 }
-fn decode_godel_to_list(g: u128) -> Vec<u128> {
-    unimplemented!();
+fn decode_godel_to_list(g: &Godel) -> Vec<Godel> {
+    if *g == 0 { return Vec::new(); }
+    let (x, xs) = decode_pair1(g);
+    let mut gs = vec![x]; 
+    gs.splice(gs.len().., decode_godel_to_list(&xs));
+    gs 
 }
```

Comment:
I had a suspicion when I saw the splice that the code was going to look fishy:
Your version:
```
example::decode_godel_to_list:
        push    rbp
        push    r15
        push    r14
        push    r13
        push    r12
        push    rbx
        sub     rsp, 152
        mov     r14, rdi
        mov     rax, rsi
        or      rax, rdx
        je      .LBB7_1
        lea     rdi, [rsp + 48]
        call    qword ptr [rip + example::decode_pair1@GOTPCREL]
        mov     r15, qword ptr [rsp + 72]
        mov     r12, qword ptr [rsp + 64]
        mov     rbp, qword ptr [rsp + 48]
        mov     r13, qword ptr [rsp + 56]
        mov     edi, 16
        mov     esi, 8
        call    qword ptr [rip + __rust_alloc@GOTPCREL]
        test    rax, rax
        je      .LBB7_22
        mov     rbx, rax
        mov     qword ptr [rax], rbp
        mov     qword ptr [rax + 8], r13
        mov     qword ptr [rsp], rax
        movaps  xmm0, xmmword ptr [rip + .LCPI7_0]
        movups  xmmword ptr [rsp + 8], xmm0
        lea     rdi, [rsp + 24]
        mov     rsi, r12
        mov     rdx, r15
        call    qword ptr [rip + example::decode_godel_to_list@GOTPCREL]
        mov     qword ptr [rsp + 16], 1
        add     rbx, 16
        mov     r12, qword ptr [rsp + 24]
        mov     r15, qword ptr [rsp + 32]
        mov     rdx, qword ptr [rsp + 40]
        mov     rbp, rdx
        shl     rbp, 4
        lea     r13, [r12 + rbp]
        movaps  xmm0, xmmword ptr [rip + .LCPI7_1]
        movaps  xmmword ptr [rsp + 80], xmm0
        mov     qword ptr [rsp + 96], rbx
        mov     qword ptr [rsp + 104], rbx
        mov     rax, rsp
        mov     qword ptr [rsp + 112], rax
        mov     qword ptr [rsp + 120], r12
        mov     qword ptr [rsp + 128], r15
        mov     qword ptr [rsp + 136], r12
        mov     qword ptr [rsp + 144], r13
        test    rdx, rdx
        je      .LBB7_6
        lea     rbx, [rsp + 120]
        mov     rdi, rsp
        mov     esi, 1
        call    alloc::raw_vec::RawVec<T,A>::reserve::do_reserve_and_handle
        mov     rcx, qword ptr [rsp + 16]
        mov     rsi, rcx
        shl     rsi, 4
        add     rsi, qword ptr [rsp]
        add     rbp, -16
        mov     eax, ebp
        shr     eax, 4
        add     eax, 1
        and     rax, 7
        je      .LBB7_11
        neg     rax
        xor     edx, edx
        mov     rdi, r12
.LBB7_13:
        movups  xmm0, xmmword ptr [rdi]
        add     rdi, 16
        movups  xmmword ptr [rsi], xmm0
        add     rsi, 16
        add     rdx, -1
        cmp     rax, rdx
        jne     .LBB7_13
        sub     rcx, rdx
        cmp     rbp, 112
        jae     .LBB7_16
        jmp     .LBB7_18
.LBB7_1:
        mov     rax, qword ptr [rip + .L__unnamed_2]
        mov     qword ptr [r14], rax
        xorps   xmm0, xmm0
        movups  xmmword ptr [r14 + 8], xmm0
        jmp     .LBB7_2
.LBB7_6:
        mov     ecx, 1
        jmp     .LBB7_18
.LBB7_11:
        mov     rdi, r12
        cmp     rbp, 112
        jb      .LBB7_18
.LBB7_16:
        xor     ebp, ebp
.LBB7_17:
        movups  xmm0, xmmword ptr [rdi + rbp]
        movups  xmmword ptr [rsi + rbp], xmm0
        movups  xmm0, xmmword ptr [rdi + rbp + 16]
        movups  xmmword ptr [rsi + rbp + 16], xmm0
        movups  xmm0, xmmword ptr [rdi + rbp + 32]
        movups  xmmword ptr [rsi + rbp + 32], xmm0
        movups  xmm0, xmmword ptr [rdi + rbp + 48]
        movups  xmmword ptr [rsi + rbp + 48], xmm0
        movups  xmm0, xmmword ptr [rdi + rbp + 64]
        movups  xmmword ptr [rsi + rbp + 64], xmm0
        movups  xmm0, xmmword ptr [rdi + rbp + 80]
        movups  xmmword ptr [rsi + rbp + 80], xmm0
        movups  xmm0, xmmword ptr [rdi + rbp + 96]
        movups  xmmword ptr [rsi + rbp + 96], xmm0
        movups  xmm0, xmmword ptr [rdi + rbp + 112]
        movups  xmmword ptr [rsi + rbp + 112], xmm0
        add     rcx, 8
        lea     rax, [rdi + rbp]
        add     rax, 128
        sub     rbp, -128
        cmp     rax, r13
        jne     .LBB7_17
.LBB7_18:
        mov     qword ptr [rsp + 16], rcx
        test    r15, r15
        je      .LBB7_21
        shl     r15, 4
        je      .LBB7_21
        mov     edx, 8
        mov     rdi, r12
        mov     rsi, r15
        call    qword ptr [rip + __rust_dealloc@GOTPCREL]
.LBB7_21:
        mov     rax, qword ptr [rsp + 16]
        mov     qword ptr [r14 + 16], rax
        movups  xmm0, xmmword ptr [rsp]
        movups  xmmword ptr [r14], xmm0
.LBB7_2:
        mov     rax, r14
        add     rsp, 152
        pop     rbx
        pop     r12
        pop     r13
        pop     r14
        pop     r15
        pop     rbp
        ret
.LBB7_22:
        mov     edi, 16
        mov     esi, 8
        call    qword ptr [rip + alloc::alloc::handle_alloc_error@GOTPCREL]
        ud2
        mov     r14, rax
        lea     rdi, [rsp + 80]
        call    core::ptr::drop_in_place<alloc::vec::drain::Drain<u128>>
        mov     rdi, rbx
        call    core::ptr::drop_in_place<alloc::vec::into_iter::IntoIter<u128>>
        jmp     .LBB7_8
        mov     r14, rax
.LBB7_8:
        mov     rdi, rsp
        call    core::ptr::drop_in_place<alloc::vec::Vec<u128>>
        mov     rdi, r14
        call    _Unwind_Resume@PLT
        ud2
```
For every vec you create, you're allocating the data, filling it, copying it into the caller's vec, then deallocating it. This turns your algorithm's time complexity from linear to polynomial.

Iterator example:
```
example::decode_godel_to_list_iterator:
        push    rbp
        push    r15
        push    r14
        push    r13
        push    r12
        push    rbx
        sub     rsp, 56
        mov     r15, qword ptr [rip + .L__unnamed_2]
        mov     qword ptr [rdi], r15
        xorps   xmm0, xmm0
        movups  xmmword ptr [rdi + 8], xmm0
        mov     rax, rsi
        or      rax, rdx
        je      .LBB5_8
        mov     qword ptr [rsp + 8], rdi
        lea     rax, [rdi + 8]
        mov     qword ptr [rsp], rax
        xor     ebx, ebx
        xor     r12d, r12d
.LBB5_2:
        lea     rdi, [rsp + 24]
        call    qword ptr [rip + example::decode_pair1@GOTPCREL]
        mov     r14, qword ptr [rsp + 24]
        mov     rbp, qword ptr [rsp + 32]
        mov     rax, qword ptr [rsp + 48]
        mov     qword ptr [rsp + 16], rax
        mov     r13, qword ptr [rsp + 40]
        cmp     r12, rbx
        mov     rdi, qword ptr [rsp + 8]
        jne     .LBB5_5
        mov     r15, rdi
        mov     rsi, rbx
        call    alloc::raw_vec::RawVec<T,A>::reserve_for_push
        mov     rax, qword ptr [r15]
        mov     r12, qword ptr [r15 + 16]
        mov     rdi, r15
        mov     r15, rax
.LBB5_5:
        mov     rax, r12
        shl     rax, 4
        mov     qword ptr [r15 + rax + 8], rbp
        mov     qword ptr [r15 + rax], r14
        add     r12, 1
        mov     qword ptr [rdi + 16], r12
        mov     rax, r13
        mov     rdx, qword ptr [rsp + 16]
        or      rax, rdx
        je      .LBB5_8
        mov     rsi, r13
        mov     rax, qword ptr [rsp]
        mov     rbx, qword ptr [rax]
        jmp     .LBB5_2
.LBB5_8:
        mov     rax, rdi
        add     rsp, 56
        pop     rbx
        pop     r12
        pop     r13
        pop     r14
        pop     r15
        pop     rbp
        ret
        mov     rbx, rax
        mov     rdi, r15
        call    core::ptr::drop_in_place<alloc::vec::Vec<u128>>
        mov     rdi, rbx
        call    _Unwind_Resume@PLT
        ud2
```

Link to full comparison: https://godbolt.org/z/bnKWbxjbW